### PR TITLE
When conditionally returning,  add condition to non returning path

### LINKED ISSUE #1182: Issue with getPureBinaryOperationResultType
The following example fails to interpret in Prepack:

```javascript
// add at runtime:var global=this;this.nativePerformanceNow = Date.now;
if (global.__assumeDataProperty) __assumeDataProperty(this, "nativePerformanceNow", function() {
  if (this.__residual)
    return __residual("number", function(global) {
      return global.nativePerformanceNow();
    }, global);
  else return this.nativePerformanceNow();
}, "SKIP_INVARIANT");
let performanceNow = nativePerformanceNow;

let timespans = {};

function addTimespan(key) {
  timespans[key] = {};
}

function startTimespan(key) {
  if (timespans[key]) return;
  timespans[key] = { startTime: performanceNow() };
}

function stopTimespan(key) {
  const timespan = timespans[key];
  if (!timespan || !timespan.startTime) return;
  timespan.endTime = performanceNow();
  // Following line has a problem
  timespan.totalTime = timespan.endTime - (timespan.startTime || 0);
}

startTimespan("hello");
stopTimespan("hello");

inspect = function() { return true; }
```

`timespan.endTime` seems to be a conditional `AbstractValue` that is `undefined | some abstract number` which is not allowed during subtraction.

### PR DESCRIPTION
**Pull request by simonhj**
Release note: Use conditional returns to refine the path condition. Fixes #1182

When one branch of a conditional is an abrupt completion add the join condition to the path condition. The previous path condition is stored in the saved completion and restored at join point.

### CODE REVIEW

File: src/methods/join.js [hunk 1]
```diff
@@ -445,10 +469,10 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects);
+      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects,[],[]);
```
Comment (hermanventer) [inline src/methods/join.js]: I expected pnc to have a merged path condition that is the inverse of c.joinCondition.

File: src/methods/join.js [hunk 2]
```diff
@@ -445,10 +469,10 @@ export class JoinImplementation {
...
       return [c.consequentEffects, pnc];
     } else if (c.alternate instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, c.consequent, c.consequentEffects, v, empty_effects);
+      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, c.consequent, c.consequentEffects, v, empty_effects, [],[]);
```
Comment (hermanventer) [inline src/methods/join.js]: merged path should be c.joinCondition.

File: src/methods/join.js [hunk 3]
```diff
@@ -572,7 +596,7 @@ export class JoinImplementation {
         savedEffects = result2.savedEffects;
       }
       invariant(value instanceof Value);
-      return new PossiblyNormalCompletion(value, joinCondition, result1, e1, result2, e2, savedEffects);
+      return new PossiblyNormalCompletion(value, joinCondition, result1, e1, result2, e2, [joinCondition], [], savedEffects);
```
Comment (hermanventer) [inline src/methods/join.js]: I expected to see a merged condition that is [!joinCondition]+result2.path.

File: src/methods/join.js [hunk 4]
```diff
@@ -582,7 +606,7 @@ export class JoinImplementation {
         savedEffects = result1.savedEffects;
       }
       invariant(value instanceof Value);
-      return new PossiblyNormalCompletion(value, joinCondition, result1, e1, result2, e2, savedEffects);
+      return new PossiblyNormalCompletion(value, joinCondition, result1, e1, result2, e2, [joinCondition], [], savedEffects);
```
Comment (hermanventer) [inline src/methods/join.js]: I expected to see [joinCondition]+result1.path

File: src/methods/join.js [hunk 5]
```diff
@@ -151,6 +151,22 @@ export class JoinImplementation {
     pnc: PossiblyNormalCompletion,
     c: PossiblyNormalCompletion
   ): PossiblyNormalCompletion {
+    //merge the two pathConditions
+    let mergedPath = [];
+    mergedPath.concat(pnc.path);
+    mergedPath.concat(c.path);
+    let savedPathConditions = []
```
[the remark this answers is not in the document]
Comment (simonhj) [reply src/methods/join.js]: Because they should always be equal right?

File: test/serializer/basic/ConditionalReturn.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+let b = global.__abstract ? _abstract("boolean", "true") : true;
```
Comment (hermanventer) [inline test/serializer/basic/ConditionalReturn.js]: Try __abstract instead of _abstract

File: src/methods/join.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hermanventer) [reply src/methods/join.js]: I don't think so. The second completion starts from a fork point inside the normal path of the first completion and presumably has its merged path as its saved path. The result, however, is a single completion with a single normal path, starting at the fork point that led to pnc, so the saved path should be the one that applied to pnc, since that is the only path that all of the threads collected in the result will have in common.

File: test/serializer/basic/ConditionalReturn.js (hunk 6 above)
Comment (simonhj) [reply test/serializer/basic/ConditionalReturn.js]: Ahh, good catch. That fixes it.

File: src/methods/join.js [hunk 7]
```diff
@@ -445,10 +458,29 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects);
+      let pathCondtion = [(AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition): any)];
```
Comment (simonhj) [inline src/methods/join.js]: Open to suggestions on this one. I don't understand why `AbstractValue.createFromUnaryOp` returns a  `Value` instead of an `AbstractValue`.

Comment (hermanventer) [reply src/methods/join.js]: Every time you create an abstract value it is simplified in the current path context. This can result in a concrete value.

File: src/methods/join.js [hunk 8]
```diff
@@ -445,10 +458,29 @@ export class JoinImplementation {
...
+      let pnc = new PossiblyNormalCompletion(
+        v,
+        c.joinCondition,
+        v,
+        empty_effects,
+        c.alternate,
+        c.alternateEffects,
+        pathCondtion,
```
Comment (hermanventer) [inline src/methods/join.js]: typo

File: src/methods/join.js (hunk 7 above)
Comment (simonhj) [reply src/methods/join.js]: I see. Makes sense. Thanks

File: src/methods/join.js [hunk 9]
```diff
@@ -445,10 +458,29 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects);
+      let pathCondition = [(AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition): any)];
```
Comment (hermanventer) [inline src/methods/join.js]: Casting to any is generally not OK and this case is no exception. First assign the result of createFromUnaryOp to a local and add it to path only if it is in fact an AbstractValue. It does not seem correct for the result to be something other than an AbstractValue, so you may want to add an invariant to that effect.

File: src/methods/join.js [hunk 10]
```diff
@@ -458,7 +458,7 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let negation = AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition)
+      let negation = AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition);
       invariant(negation instanceof AbstractValue);
```
Comment (hermanventer) [inline src/methods/join.js]: It is always a good idea to add a comment to document why you are convinced that the invariant will always be true.

File: src/completions.js [hunk 11]
```diff
@@ -106,6 +106,8 @@ export class PossiblyNormalCompletion extends NormalCompletion {
     consequentEffects: Effects,
     alternate: Completion | Value,
     alternateEffects: Effects,
+    pathCondtion: Array<AbstractValue>,
```
Comment (hermanventer) [inline src/completions.js]: pathCondtion -> pathConditions

File: src/methods/join.js [hunk 12]
```diff
@@ -151,6 +151,11 @@ export class JoinImplementation {
     pnc: PossiblyNormalCompletion,
     c: PossiblyNormalCompletion
   ): PossiblyNormalCompletion {
+    //merge the two pathConditions
+    let mergedPath = [];
```
Comment (hermanventer) [inline src/methods/join.js]: Rather call this composedPath.

File: src/methods/join.js [hunk 13]
```diff
@@ -151,6 +151,11 @@ export class JoinImplementation {
...
+    mergedPath = mergedPath.concat(pnc.pathCondtion);
+    mergedPath = mergedPath.concat(c.pathCondtion);
```
Comment (hermanventer) [inline src/methods/join.js]: You can shorten this to:

let mergedPath = pnc.pathConditions.concat(c.pathConditions);

File: src/methods/join.js [hunk 14]
```diff
@@ -445,10 +458,31 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects);
+      let negation = AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition);
+      invariant(negation instanceof AbstractValue);
+      let pathCondition = [negation];
```
Comment (hermanventer) [inline src/methods/join.js]: c.alternate may itself be a PossiblyNormalCompletion and, if so, you should append its pathConditions to get the path that applies to the normal branch.

File: src/methods/join.js [hunk 15]
```diff
@@ -445,10 +458,31 @@ export class JoinImplementation {
...
     } else if (c.alternate instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, c.consequent, c.consequentEffects, v, empty_effects);
+      let pnc = new PossiblyNormalCompletion(
+        v,
+        c.joinCondition,
+        c.consequent,
+        c.consequentEffects,
+        v,
+        empty_effects,
+        [c.joinCondition],
```
Comment (hermanventer) [inline src/methods/join.js]: You need to account for the possibility that c.consequent may be a PossiblyNormalCompletion.

File: src/realm.js [hunk 16]
```diff
@@ -493,10 +493,16 @@ export class Realm {
   composeWithSavedCompletion(completion: PossiblyNormalCompletion): Value {
     if (this.savedCompletion === undefined) {
       this.savedCompletion = completion;
+      this.savedCompletion.savedPathConditions = this.pathConditions;
       this.captureEffects(completion);
     } else {
       this.savedCompletion = Join.composePossiblyNormalCompletions(this, this.savedCompletion, completion);
     }
+    if (completion.consequent instanceof AbruptCompletion) {
+      Path.pushInverseAndRefine(completion.joinCondition);
```
Comment (hermanventer) [inline src/realm.js]: If completion.alternate is a PossiblyNormalCondition, you need to push all of its pathConditions.

File: src/realm.js [hunk 17]
```diff
@@ -493,10 +493,16 @@ export class Realm {
...
+    } else if (completion.alternate instanceof AbruptCompletion) {
+      Path.pushAndRefine(completion.joinCondition);
```
Comment (hermanventer) [inline src/realm.js]: Need to handle the case where completion.consequent is a PossiblyNormalCondition.

File: src/methods/join.js (hunk 14 above)
Comment (simonhj) [reply src/methods/join.js]: How so? JoinedAbruptCompletions has two AbruptCompletions as children, they cannot be PossibleNormal can they?

Comment (hermanventer) [reply src/methods/join.js]: Ah yes, you are quite right. I take that back.

**Comment by simonhj** [comment]
Agreed on testing. Any ideas on good corner cases to cover?

File: src/completions.js [hunk 18]
```diff
@@ -106,7 +106,7 @@ export class PossiblyNormalCompletion extends NormalCompletion {
     consequentEffects: Effects,
     alternate: Completion | Value,
     alternateEffects: Effects,
-    pathCondtion: Array<AbstractValue>,
+    pathCondtions: Array<AbstractValue>,
```
Comment (hermanventer) [inline src/completions.js]: pathConditions

**Comment by simonhj** [comment]
@hermanventer do you have a good idea for a testcase testing "restore previous path condition" part of this functionality?

**Comment by hermanventer** [comment]
For testing I look to code coverage data for inspiration. If a particular branch is not taken, I try to construct a case that will take that branch. The previous path condition needs to be restored when the normal branch of a fork joins up with the abrupt branches. Following such a join point, all we know about the path is what we knew before the fork occurred. Of course, path restoration needs to take into account any information discovered since the fork (I.e. the joined effects).

**Comment by hermanventer** [comment]
Please update your initial comment with a release note and issue number. You can also remove the unfortunate paragraph.
